### ISSUE TITLE
pgtype.TextArray for IN clause parameter with multiple columns giving error "ERROR: input of anonymous composite types is not implemented (SQLSTATE 0A000)"

### ISSUE THREAD

**Issue by pranay-1995** [jackc/pgtype#157]
stmt, err := r.DB.Prepare("SELECT * FROM emp WHERE (name, code) IN ($1)")
inClause := [][]string{}{{"emp1", "123"}, {"emp2", "456"}}
var tarray pgtype.TextArray
tarray.Set(inClause)
rows, err := stmt.Query(inClause)

Here, error is returned saying `ERROR: input of anonymous composite types is not implemented (SQLSTATE 0A000)`

Expected - Query should have got executed successfully with IN clause.

**Comment by jackc** [comment]
I believe that is a PostgreSQL limitation. I don't think that can work even in psql.

**Comment by pranay-1995** [comment]
@jackc The query that I mentioned above, i.e., `SELECT * FROM emp WHERE (name, code) IN (("emp1", "123"), ("emp2", "456"))` works in PostgreSQL. 
But I couldn't find any way to form this query with prepared statement using pgtype. What I suspect is that pgtype.TextArray adds single quotes around the in clause parameter in case of 2d-array. So the effective query formed becomes,
SELECT * FROM emp WHERE (name, code) IN ('("emp1", "123"), ("emp2", "456")')`.
When I try above query directly on PostgreSQL, it also gives the same error as mentioned above. Shouldn't pgtype bind it correctly without single quotes?

**Comment by jackc** [comment]
> The query that I mentioned above, i.e., SELECT * FROM emp WHERE (name, code) IN (("emp1", "123"), ("emp2", "456")) works in PostgreSQL.

That is a very different query from `SELECT * FROM emp WHERE (name, code) IN ($1)`.

Use psql and try using `prepare` and `execute`. If you can get it to work I would like to know how.

**Comment by pranay-1995** [comment]
@jackc I am able to use IN clause with multiple columns in Java/Kotlin successfully. There also I am pointing to PostgreSQL. So isn't it safe to say that it is not PostgresSQL limitation?

Java/Kotlin code,

```
@Query(
        """
            DELETE FROM emp
            WHERE (name, code) IN (:emps)
            RETURNING name, code
        """
    )
    fun deleteEmps(
        emps: List<Array<Any>>
    ): Flux<Emp>
```

**Comment by jackc** [comment]
> So isn't it safe to say that it is not PostgresSQL limitation?

No. Your Java / Kotlin code is apparently doing some query rewriting. That's why I suggested using `psql` directly so you could experiment with what PostgreSQL can do.

An `IN` clause can contain multiple elements but it does not automatically expand a single array into multiple elements. 

For example, you can do this:

```sql
select * from users where id in ($1, $2, $3)
```

when each of `$1`, `$2`, and `$3` are the type of `users.id` but you cannot do this

```sql
select * from users where id in ($1)
```

when `$1` is an array of the type of `users.id`.

In addition, even if that was worked around, there is still the `ERROR: input of anonymous composite types is not implemented (SQLSTATE 0A000)` error. As it says, PostgreSQL does not allow the input of anonymous composite types.

What is actually happening when you are using an ORM that makes this work is it is rewriting the query. When you do something like:

```sql
SELECT * FROM emp WHERE (name, code) IN (:emps)
```

Your ORM is actually rewriting the SQL to something like:

```sql
SELECT * FROM emp WHERE (name, code) IN (($1, $2), ($3, $4))
```

pgx does not do this rewriting. It is sending your query directly to the PostgreSQL server as a prepared / parameterized statement. Again, use `prepare` and `execute` in `psql` to see what PostgreSQL can do directly.

**Comment by pranay-1995** [comment]
So is there any way to have prepared statement with IN clause with multiple columns using pgx/pgtype?

**Comment by jackc** [comment]
> So is there any way to have prepared statement with IN clause with multiple columns using pgx/pgtype?

You'd have to write the query with the parts separated. e.g.

```sql
SELECT * FROM emp WHERE (name, code) IN (($1, $2), ($3, $4))
```

**Comment by pranay-1995** [comment]
Well, that might not feasible considering array can be large and we will need binding for each parameter, which can make code not much readable. I think we should have some more generic way in pgx/pgtype to handle such IN clause.

I have found a workaround for now but not sure if its efficient one,

```
stmt, err := r.DB.Prepare("SELECT * FROM emp WHERE (name, code) IN (SELECT * FROM json_each_text(json_object($1::text[])))")
inClause := [][]string{}{{"emp1", "123"}, {"emp2", "456"}}
var tarray pgtype.TextArray
tarray.Set(inClause)
rows, err := stmt.Query(tarray)
```
